### DISCUSSION TITLE
Silencing metrics / Skip publishing metrics based on an environment variable

### DISCUSSION THREAD

**Discussion by zirkelc** [aws-powertools/powertools-lambda-typescript#3315] [Q&A] [answered]
Hi,

I wanted to know if it is possible to _silence metrics_ based on an environment variable in the same way it is possible to [silence logs](https://docs.powertools.aws.dev/lambda/typescript/main/core/logger/#silencing-logs).

Basically, I want to omit publishing metrics if the environment **is not** production. Ideally, that should be controllable with an env var.

**Comment by dreamorosi** [comment]
Hi @zirkelc, at the moment this is not possible.

We have equivalent settings for Logger (as you pointed out), Tracer, and Idempotency - basically all the other utilities with side effects - but this feature request has never come up.

I think we could either add a `POWERTOOLS_METRICS_DISABLED` environment variable (name TBD) or add this to the [existing `POWERTOOLS_DEV` env variable](https://docs.powertools.aws.dev/lambda/typescript/latest/#environment-variables).

For consistency & ease of usage we should probably add both. This would be similar to what already happens for Tracer.

This is something we should be able to add relatively quickly, even for the next release. I just would like to allow one or two days max for other maintainers to give an opinion on the name of the environment variable, since it'll be a feature that will likely be adopted by other Powertools for AWS Lambda runtimes.

cc @aws-powertools/lambda-python-core, @aws-powertools/lambda-dotnet-core

**Reply by dreamorosi** [reply]
So it looks like we already had a request for this feature and the discussion ended at the same conclusion of the proposal above: aws-powertools/powertools-lambda-python#2996

**Reply by zirkelc** [reply]
Thanks for pointing this out, I hadn't thought to check the other repos.  I'm not sure if there is a consensus on this issue, but I infer from the labels that a PR would be accepted? I could submit one even if it was written in TypeScript and not Python.

**Reply by dreamorosi** [reply]
Thanks for your patience, I have brought this up with the team during our internal weekly planning meeting and we have agreed to move forward with the feature, so a PR would be welcome.

I'm converting this to an issue so we can track the item. In terms of implementation I think the easiest and less invasive way of implementing this would be to disable the utility by making the `Metrics.publishStoredMetrics()` a no-op when certain conditions are met.

Specifically:
- by default, Metrics is enabled (current behavior)
- if the `POWERTOOLS_METRICS_DISABLED` env variable is present and [truthy](https://github.com/aws-powertools/powertools-lambda-typescript/blob/5350afed92d02b7b8d47f387705f70c59deeeb65/packages/commons/src/config/EnvironmentVariablesService.ts#L90-L94) when the `Metrics` class is instantiated, then it becomes disabled
- if dev mode is enabled by means of `POWERTOOLS_DEV` which can be checked with [this method](https://github.com/aws-powertools/powertools-lambda-typescript/blob/5350afed92d02b7b8d47f387705f70c59deeeb65/packages/commons/src/config/EnvironmentVariablesService.ts#L81-L83), when the `Metrics` class is instantiated, then it also becomes disabled
- If `POWERTOOLS_METRICS_DISABLED` is set to `false`, and dev mode is enabled, then `POWERTOOLS_METRICS_DISABLED` takes precedence and the utility stays enabled.

This last point, while slightly counter intuitive, provides an escape hatch for those who want to use `POWERTOOLS_DEV` for other utilities/features but want to keep Metrics enabled.

--

If you have any questions during the implementation, please let me know.

**Reply by leandrodamascena** [reply]
I'll wait for TypeScript to be implemented before moving forward with Python.

Thanks for working on this @zirkelc.

**Reply by dreamorosi** [reply]
I've opened the issue (#3329), whenever you start working on it please leave a comment there so I can assign it to you for tracking purposes - GitHub doesn't allow me to assign issues to members of the community who haven't interacted with an issue first.

Also we are planning a release on Thursday end of the day / Friday morning, just mentioning it to manage expectations. If the PR is in by Wednesday end of the day, we'll do our best to review it and merge it before the release - otherwise it'll have to wait for the next one which will likely be the first week of December.

**Reply by zirkelc** [reply]
I added a comment so you can assign it to me.

**Comment by dreamorosi** [comment, marked as answer]
The change has been merged in #3351 and will be included in the next release.

Thank you again @zirkelc for bringing this up and for the contribution!
